### PR TITLE
syncer(dm): support GBK replication

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
close https://github.com/pingcap/tiflow/issues/3774

### What is changed and how it works?
- use `_binary'x'` to insert GBK data from binlog, and add simple test.
- For DDL, decode the query when it's GBK

Also move some function logic together

### Check List 

Tests 

 - Integration test

Code changes

Side effects

 - Breaking backward compatibility

Related changes

### Release note 

```release-note
Support replicate GBK tables as-is
```

### CODE REVIEW

**Comment by lance6716** [comment]
waiting nightly TiDB build and https://github.com/pingcap/tidb-tools/issues/572

File: dm/syncer/syncer.go [hunk 1]
```diff
@@ -2333,6 +2333,19 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
 		return nil
 	}
 
+	codec, err := event.GetCharsetCodecByStatusVars(ev.StatusVars)
+	if err != nil {
+		s.tctx.L().Error("get charset codec failed, will treat query as utf8", zap.Error(err))
+	}
+	if codec != nil {
```
Comment (lichunzhu) [inline dm/syncer/syncer.go:2340]: weakly suggestion
```suggestion
	} else codec != nil {
```

File: dm/tests/_utils/test_prepare [hunk 2]
```diff
@@ -219,7 +219,9 @@ function run_sql_tidb() {
 function run_sql_tidb_with_retry() {
     rc=0
     for ((k=1; k<11; k++)); do
+        set +e
```
Comment (lichunzhu) [inline dm/tests/_utils/test_prepare:222]: Why? Leave a comment here?

File: dm/tests/gbk/data/db1.increment.sql [hunk 3]
```diff
@@ -0,0 +1,10 @@
+use `gbk`;
```
Comment (lichunzhu) [inline dm/tests/gbk/data/db1.increment.sql:1]: Do we need a test for a single gbk charset column?
```sql
create table t20 (id int, name varchar(20) character set gbk, primary key(`id`)) character set utf8;
```

**Review by lichunzhu — APPROVED** [review]
Rest LGTM

File: dm/pkg/binlog/event/util.go [hunk 4]
```diff
@@ -376,6 +379,35 @@ func GetServerCollationByStatusVars(statusVars []byte, idAndCollationMap map[int
 	return idAndCollationMap[int(v)], err
 }
 
+// GetCharsetCodecByStatusVars returns an encoding.Encoding to encode and decode original query if needed.
+func GetCharsetCodecByStatusVars(statusVars []byte) (encoding.Encoding, error) {
```
Comment (Ehco1996) [inline dm/pkg/binlog/event/util.go:383]: this func called `GetCharsetCodecByStatusVars` but only return chartset code when chartset is gbk , this looks litte strange, maybe change a name?

**Review by Ehco1996 — COMMENTED** [review]
rest LGTM

File: dm/pkg/binlog/event/util.go (hunk 4 above)
Comment (lance6716) [reply dm/pkg/binlog/event/util.go:383]: the comment says "returns ... if needed", we can return more codec if user wants

Comment (Ehco1996) [reply dm/pkg/binlog/event/util.go:383]: ok i see

**Review by Ehco1996 — APPROVED** [review]

**Pull request merged**